lib/beautiful/colorful.lua

```diff
@@ -0,0 +1,1395 @@
+    local green = clamp(1 - self.green, 0, 1) * 255
+    local blue = clamp(1 - self.blue, 0, 1) * 255
+    return color.from.rgba(red, green, blue, self.alpha * 255)
+end
+
+-- }}} RGB-adjusting functions
+
+-- {{{ HSL-adjusting functions
+
+--- Returns a color darkened by a specified amount
+-- @tparam number v A number [0,1] of how much to decrease Lightness
+-- @return A new color object representing the darkened color, or nil if invalid
+-- @usage local darker_color = c:darken(0.25)
+function color:darken(v)
+    local lightness = clamp(self.lightness - v, 0, 1)
+    return color.from.hsla(self.hue, self.saturation, lightness, self.alpha)
+end
+
+--- Returns a color lightened by a specified amount
+-- @tparam number v A number [0,1] of how much to increase Lightness
+-- @return A new color object representing the lightened color, or nil if invalid
+-- @usage local lighter_color = c:lighten(0.25)
+function color:lighten(v)
+    local lightness = clamp(self.lightness + v, 0, 1)
+    return color.from.hsla(self.hue, self.saturation, lightness, self.alpha)
+end
+
+--- Returns a color saturated by a specified amount
+-- @tparam number v A number [0,1] of how much to increase Saturation
+-- @return A new color object representing the saturated color, or nil if invalid
+-- @usage local saturated_color = c:saturate(0.25)
+function color:saturate(v)
+    local saturation = clamp(self.saturation + v, 0, 1)
+    return color.from.hsla(self.hue, saturation, self.lightness, self.alpha)
+end
+
+--- Returns a color desaturated by a specified amount
+-- @tparam number v A number [0,1] of how much to decrease Saturation
+-- @return A new color object representing the desaturated color, or nil if invalid
+-- @usage local desaturated_color = c:desaturate(0.25)
+function color:desaturate(v)
+    local saturation = clamp(self.saturation - v, 0, 1)
+    return color.from.hsla(self.hue, saturation, self.lightness, self.alpha)
+end
+
+--- Returns the grayscale of a color
+-- Reduces Saturation to 0
+-- @return A new color object representing the grayscaled color, or nil if invalid
+-- @usage local grayscaled_color = c:grayscale()
+function color:grayscale()
+    return color.from.hsla(self.hue, 0, self.lightness, self.alpha)
+end
+
+-- }}} HSL-adjusting functions
+
+-- {{{ Color harmony functions
+
+--- Returns the complement of a color
+-- Adds 180 to Hue (hue will self-adjust to be within [0,360])
+-- @return A new color object representing the complementary color, or nil if invalid
+-- @usage local complemented_color = c:complement()
+function color:complement()
+    local hue = self.hue + 180
+    return color.from.hsla(hue, self.saturation, self.lightness, self.alpha)
+end
+
+--- Returns the left and right analogous colors (30deg hue shift)
+-- @return A new color object representing a -30 degree hue shift of the color, or nil if invalid
+-- @return A new color object representing a +30 degree hue shift of the color, or nil if invalid
```

Comment:
Wait, why `invalid`? Isn't hue an angle that "wraps around"?

Reply:
If somehow the color creation goes wrong then it's a fallback. That's what it means.